File: src/services/pronunciationService.js
import { AppStrings } from '../core/constants/appStrings';

/**
 * Evaluate pronunciation by sending audio to the API
 * @param {string} targetText - Expected text
 * @param {Blob} audioBlob - Recorded audio blob
 * @returns {Promise<{targetText: string, recognizedText: string, accuracy: number, passed: boolean}>}
 */
export async function evaluate(targetText, audioBlob) {
  const formData = new FormData();
  formData.append('target_text', targetText);
  formData.append('audio_file', audioBlob, 'recording.wav');

  const response = await fetch(AppStrings.evalEndpoint, {
    method: 'POST',
    body: formData,
  });

  if (!response.ok) {
    throw new Error(`API Error: ${response.status}`);
  }

  const json = await response.json();
  const data = json.data || json;

  const target = (data.target_text || targetText).toLowerCase();
  const recognized = (data.recognized_text || '').toLowerCase();

  // Calculate accuracy word-by-word
  const targetWords = target.split(/\s+/).filter(Boolean);
  const recognizedWords = recognized.split(/\s+/).filter(Boolean);

  let correct = 0;
  for (let i = 0; i < targetWords.length; i++) {
    if (i < recognizedWords.length && targetWords[i] === recognizedWords[i]) {
      correct++;
    }
  }

  const accuracy = targetWords.length > 0
    ? Math.round((correct / targetWords.length) * 100)
    : 0;

  return {
    targetText: target,
    recognizedText: recognized,
    accuracy,
    passed: accuracy >= 80,
  };
}

/**
 * Get TTS audio URL for given text
 * @param {string} text
 * @returns {string}
 */
export function getTtsUrl(text) {
  return `${AppStrings.ttsEndpoint}?text=${encodeURIComponent(text)}`;
}

File: lib/services/pronunciation_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../core/constants/app_strings.dart';

/// Kết quả chấm phát âm từ API.
class PronunciationResult {
  final String targetText;
  final String recognizedText;
  final int accuracy;
  final bool passed;

  PronunciationResult({
    required this.targetText,
    required this.recognizedText,
    required this.accuracy,
    required this.passed,
  });
}

/// Service dùng chung cho cả Practice và Exam.
/// Gọi API evaluate phát âm và tính accuracy.
class PronunciationService {
  PronunciationService._();

  static const int passThreshold = 80;

  /// Gửi file audio lên API, trả về [PronunciationResult].
  /// Throws [Exception] nếu lỗi mạng hoặc server.
  static Future<PronunciationResult> evaluate({
    required String targetText,
    required String audioFilePath,
  }) async {
    try {
      final req = http.MultipartRequest(
        'POST',
        Uri.parse(AppStrings.evalEndpoint),
      );
      req.fields['target_text'] = targetText;
      req.files.add(
        await http.MultipartFile.fromPath('audio_file', audioFilePath),
      );

      final res = await req.send().timeout(const Duration(seconds: 30));
      final body = await res.stream.bytesToString();

      if (res.statusCode != 200) {
        throw Exception('Server trả về lỗi: ${res.statusCode}');
      }

      final data = jsonDecode(body)['data'];
      final tw = data['target_text'].toString().toLowerCase().split(' ');
      final rw = data['recognized_text'].toString().toLowerCase().split(' ');

      int correct = 0;
      for (int i = 0; i < rw.length; i++) {
        if (i < tw.length && rw[i] == tw[i]) correct++;
      }
      final acc =
          tw.isEmpty ? 0 : ((correct / tw.length) * 100).round().clamp(0, 100);

      return PronunciationResult(
        targetText: data['target_text'] ?? targetText,
        recognizedText: data['recognized_text'] ?? '',
        accuracy: acc,
        passed: acc >= passThreshold,
      );
    } catch (e) {
      throw Exception('Không thể chấm phát âm: $e');
    }
  }
}
